ui/components/src/EChart.tsx

```diff
@@ -121,62 +102,80 @@ export type OnEventsType<T> = {
 export interface EChartsProps<T> {
   option: EChartsCoreOption;
   sx?: SxProps<Theme>;
-  onChartInitialized?: (instance: ECharts) => void;
   onEvents?: OnEventsType<T>;
   _instance?: React.MutableRefObject<ECharts | undefined>;
-  parentContainer?: HTMLDivElement | null;
+  onChartInitialized?: (instance: ECharts) => void;
 }
 
-export function ECharts<T>({ sx, _instance, onChartInitialized, option, onEvents }: EChartsProps<T>) {
-  const [containerRef, setContainerRef] = useState<HTMLDivElement | null>(null);
-  const [echart, setChart] = useState<ECharts | undefined>(undefined);
-  // Create a chart instance in the container
-  useLayoutEffect(() => {
-    if (containerRef === null) return;
-
-    const chart = init(containerRef);
-    setChart(chart);
-    if (_instance !== undefined) {
-      _instance.current = chart;
+export const EChart = React.memo(function EChart<T>({
+  option,
+  sx,
+  onEvents,
+  _instance,
+  onChartInitialized,
+}: EChartsProps<T>) {
+  const prevOption = useRef<EChartsCoreOption | undefined>();
+  const containerRef = useRef<HTMLDivElement | null>(null);
+  const chartElement = useRef<ECharts | null>(null);
+
+  const initChartDom = useCallback(() => {
+    if (containerRef.current !== null && chartElement.current === null) {
+      chartElement.current = init(containerRef.current);
+      onChartInitialized?.(chartElement.current);
+      if (_instance !== undefined) {
+        _instance.current = chartElement.current;
+      }
     }
-    bindEvents(chart, onEvents);
+  }, [_instance, onChartInitialized]);
 
-    onChartInitialized?.(chart);
-
-    return () => {
-      chart.dispose();
-    };
-  }, [containerRef, _instance, onChartInitialized, onEvents]);
+  useEffect(() => {
+    const chart = chartElement.current;
+    if (chart === null || onEvents === undefined) return;
+    bindEvents(chart, onEvents);
```

Comment:
Shouldn't this be cleaning up event handlers also in a dispose since it's the effect that's binding them? Not sure if that's a thing since I'm not super familiar with the echarts API.

Reply:
We're definitely in undocumented territory, but I have [an approach now](https://github.com/perses/perses/blob/fcc5354c64e22a7cb90fcc743dc54a0c396c6888/ui/components/src/EChart.tsx#L160) that will work for unbinding events individually!